Bug — `RefundController` and `VerificationController` Import from Non-Existent Paths

## Description

Both `src/payments/refunds/refund.controller.ts` and `src/tickets/verification/verification.controller.ts` import `RolesGuard` and the `Roles` decorator from paths that do not exist:

```ts
import { RolesGuard } from '../../auth/guards/roles.guard';
import { Roles } from '../../auth/decorators/roles.decorator';
```

The directory `src/auth/guards/` only contains `jwt-auth.guard.ts`. There is no `roles.guard.ts` inside it, and `src/auth/decorators/` does not exist at all. These imports cause **compile-time failures** for both controllers.

The actual `RolesGuard` and `Roles` decorator live in `src/common/guards/roles.guard.ts` and `src/common/decorators/roles.decorator.ts` respectively.

## Steps to Reproduce

1. Run `npm run build`
2. TypeScript compiler throws: `Cannot find module '../../auth/guards/roles.guard'`

## Expected Behaviour

Both controllers should compile successfully, resolving their guards and decorators from the correct paths.

## Actual Behaviour

Build fails with module-not-found errors for both controllers.

## Proposed Fix

Update the import paths in both files to point to the `common` directory:

```ts
// src/payments/refunds/refund.controller.ts
import { RolesGuard } from '../../common/guards/roles.guard';
import { Roles } from '../../common/decorators/roles.decorator';

// src/tickets/verification/verification.controller.ts
import { RolesGuard } from '../../common/guards/roles.guard';
import { Roles } from '../../common/decorators/roles.decorator';
```

## Affected Files

- `src/payments/refunds/refund.controller.ts`
- `src/tickets/verification/verification.controller.ts`

## Complexity

**Medium**

--- a/backend/.env.example
+++ b/backend/.env.example
@@ -1,3 +1,14 @@
+# ─── Ticket Signing Keypair ────────────────────────────────────────────────
+# These keys are used ONLY for signing ticket payloads (e.g., ticketId) for gate verification.
+# They are NOT used for any on-chain transaction or asset transfer.
+TICKET_SIGNING_SECRET="SXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
+TICKET_SIGNING_PUBLIC_KEY="GXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
+# ─── SMTP Mailer Configuration ─────────────────────────────────────────────
+SMTP_HOST=smtp.example.com
+SMTP_PORT=587
+SMTP_USER=no-reply@example.com
+SMTP_PASS=your-smtp-password
+MAIL_FROM="Lumentix no-reply@example.com"
 NODE_ENV=development
 
 DB_HOST=localhost

--- a/backend/src/app.module.ts
+++ b/backend/src/app.module.ts
@@ -20,6 +20,8 @@ import { NotificationModule } from './notifications/notification.module';
 import { CurrenciesModule } from './currencies/currencies.module';
 import { ExchangeRatesModule } from './exchange-rates/exchange-rates.module';
 import { TransactionsModule } from './transactions/transactions.module';
+import { TicketsModule } from './tickets/tickets.module';
+import { AdminModule } from './admin/admin.module';
 
 @Module({
   imports: [
@@ -84,6 +86,8 @@ import { TransactionsModule } from './transactions/transactions.module';
     TransactionsModule,
     CurrenciesModule,
     ExchangeRatesModule,
+    TicketsModule,
+    AdminModule,
   ],
   controllers: [AppController],
   providers: [

--- a/backend/src/currencies/currencies.controller.ts
+++ b/backend/src/currencies/currencies.controller.ts
@@ -1,4 +1,12 @@
-import { Controller, Get, Post, Body, Patch, Param, Delete } from '@nestjs/common';
+import {
+  Controller,
+  Get,
+  Post,
+  Body,
+  Patch,
+  Param,
+  Delete,
+} from '@nestjs/common';
 import { CurrenciesService } from './currencies.service';
 import { CreateCurrencyDto } from './dto/create-currency.dto';
 import { UpdateCurrencyDto } from './dto/update-currency.dto';
@@ -19,16 +27,19 @@ export class CurrenciesController {
 
   @Get(':id')
   findOne(@Param('id') id: string) {
-    return this.currenciesService.findOne(+id);
+    return this.currenciesService.findOne(id);
   }
 
   @Patch(':id')
-  update(@Param('id') id: string, @Body() updateCurrencyDto: UpdateCurrencyDto) {
-    return this.currenciesService.update(+id, updateCurrencyDto);
+  update(
+    @Param('id') id: string,
+    @Body() updateCurrencyDto: UpdateCurrencyDto,
+  ) {
+    return this.currenciesService.update(id, updateCurrencyDto);
   }
 
   @Delete(':id')
   remove(@Param('id') id: string) {
-    return this.currenciesService.remove(+id);
+    return this.currenciesService.remove(id);
   }
 }

--- a/backend/src/currencies/currencies.service.ts
+++ b/backend/src/currencies/currencies.service.ts
@@ -2,7 +2,7 @@ import { Injectable } from '@nestjs/common';
 import { InjectRepository } from '@nestjs/typeorm';
 import { In, Repository } from 'typeorm';
 import { Currency } from './entities/currency.entity';
-
+// ...existing code...
 export interface CurrencyMeta {
   code: string;
   symbol: string;
@@ -34,5 +34,30 @@ export class CurrenciesService {
         { code: c.code, symbol: c.symbol, displayName: c.displayName },
       ]),
     );
+  }
+  async create(createCurrencyDto: Partial<Currency>): Promise<Currency> {
+    const currency = this.currencyRepository.create(createCurrencyDto);
+    return await this.currencyRepository.save(currency);
+  }
+
+  async findAll(): Promise<Currency[]> {
+    return await this.currencyRepository.find();
+  }
+
+  async findOne(id: string): Promise<Currency | undefined> {
+    const result = await this.currencyRepository.findOne({ where: { id } });
+    return result ?? undefined;
+  }
+
+  async update(
+    id: string,
+    updateCurrencyDto: Partial<Currency>,
+  ): Promise<Currency | undefined> {
+    await this.currencyRepository.update(id, updateCurrencyDto);
+    return await this.findOne(id);
+  }
+
+  async remove(id: string): Promise<void> {
+    await this.currencyRepository.delete(id);
   }
 }

--- a/backend/src/exchange-rates/exchange-rates.controller.ts
+++ b/backend/src/exchange-rates/exchange-rates.controller.ts
@@ -1,4 +1,12 @@
-import { Controller, Get, Post, Body, Patch, Param, Delete } from '@nestjs/common';
+import {
+  Controller,
+  Get,
+  Post,
+  Body,
+  Patch,
+  Param,
+  Delete,
+} from '@nestjs/common';
 import { ExchangeRatesService } from './exchange-rates.service';
 import { CreateExchangeRateDto } from './dto/create-exchange-rate.dto';
 import { UpdateExchangeRateDto } from './dto/update-exchange-rate.dto';
@@ -19,16 +27,19 @@ export class ExchangeRatesController {
 
   @Get(':id')
   findOne(@Param('id') id: string) {
-    return this.exchangeRatesService.findOne(+id);
+    return this.exchangeRatesService.findOne(id);
   }
 
   @Patch(':id')
-  update(@Param('id') id: string, @Body() updateExchangeRateDto: UpdateExchangeRateDto) {
-    return this.exchangeRatesService.update(+id, updateExchangeRateDto);
+  update(
+    @Param('id') id: string,
+    @Body() updateExchangeRateDto: UpdateExchangeRateDto,
+  ) {
+    return this.exchangeRatesService.update(id, updateExchangeRateDto);
   }
 
   @Delete(':id')
   remove(@Param('id') id: string) {
-    return this.exchangeRatesService.remove(+id);
+    return this.exchangeRatesService.remove(id);
   }
 }

--- a/backend/src/exchange-rates/exchange-rates.service.ts
+++ b/backend/src/exchange-rates/exchange-rates.service.ts
@@ -1,3 +1,5 @@
+// ...existing code...
+// ...existing code...
 import { Injectable, Logger, NotFoundException } from '@nestjs/common';
 import { InjectRepository } from '@nestjs/typeorm';
 import { Repository } from 'typeorm';
@@ -87,5 +89,32 @@ export class ExchangeRatesService {
         `Could not retrieve exchange rate for ${fromCode}→${toCode}.`,
       );
     }
+  }
+  async create(
+    createExchangeRateDto: Partial<ExchangeRate>,
+  ): Promise<ExchangeRate> {
+    const rate = this.ratesRepository.create(createExchangeRateDto);
+    return await this.ratesRepository.save(rate);
+  }
+
+  async findAll(): Promise<ExchangeRate[]> {
+    return await this.ratesRepository.find();
+  }
+
+  async findOne(id: string): Promise<ExchangeRate | undefined> {
+    const result = await this.ratesRepository.findOne({ where: { id } });
+    return result ?? undefined;
+  }
+
+  async update(
+    id: string,
+    updateExchangeRateDto: Partial<ExchangeRate>,
+  ): Promise<ExchangeRate | undefined> {
+    await this.ratesRepository.update(id, updateExchangeRateDto);
+    return await this.findOne(id);
+  }
+
+  async remove(id: string): Promise<void> {
+    await this.ratesRepository.delete(id);
   }
 }

--- a/backend/src/mailer/mailer.module.ts
+++ b/backend/src/mailer/mailer.module.ts
@@ -1,0 +1,10 @@
+import { Module } from '@nestjs/common';
+import { MailerService } from './mailer.service';
+import { ConfigModule } from '@nestjs/config';
+
+@Module({
+  imports: [ConfigModule],
+  providers: [MailerService],
+  exports: [MailerService],
+})
+export class MailerModule {}

--- a/backend/src/mailer/mailer.service.ts
+++ b/backend/src/mailer/mailer.service.ts
@@ -1,0 +1,38 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import * as nodemailer from 'nodemailer';
+
+@Injectable()
+export class MailerService {
+  private readonly logger = new Logger(MailerService.name);
+  private readonly transporter: nodemailer.Transporter;
+  private readonly mailFrom: string;
+
+  constructor(private readonly configService: ConfigService) {
+    const host = this.configService.get<string>('SMTP_HOST');
+    const port = this.configService.get<number>('SMTP_PORT');
+    const user = this.configService.get<string>('SMTP_USER');
+    const pass = this.configService.get<string>('SMTP_PASS');
+    this.mailFrom = this.configService.get<string>('MAIL_FROM') ?? '';
+
+    if (!host || !port || !user || !pass || !this.mailFrom) {
+      this.logger.warn('Missing SMTP configuration environment variables.');
+    }
+
+    this.transporter = nodemailer.createTransport({
+      host,
+      port,
+      auth: { user, pass },
+      secure: false,
+    });
+  }
+
+  async send(to: string, subject: string, html: string): Promise<void> {
+    await this.transporter.sendMail({
+      from: this.mailFrom,
+      to,
+      subject,
+      html,
+    });
+  }
+}

--- a/backend/src/notifications/notification.module.ts
+++ b/backend/src/notifications/notification.module.ts
@@ -2,12 +2,14 @@ import { Module } from '@nestjs/common';
 import { BullModule } from '@nestjs/bull';
 import { NotificationService } from './notification.service';
 import { NotificationProcessor } from './notification.processor';
+import { MailerModule } from '../mailer/mailer.module';
 
 @Module({
   imports: [
     BullModule.registerQueue({
       name: 'notifications', // Must match the string in @Processor
     }),
+    MailerModule,
   ],
   providers: [NotificationService, NotificationProcessor],
   exports: [NotificationService], // Allow Payments/Sponsors to import this

--- a/backend/src/notifications/notification.processor.ts
+++ b/backend/src/notifications/notification.processor.ts
@@ -7,38 +7,75 @@ import {
 } from '@nestjs/bull';
 import { Job } from 'bull';
 import { Logger } from '@nestjs/common';
+import { MailerService } from '../mailer/mailer.service';
 
 @Processor('notifications')
 export class NotificationProcessor {
   private readonly logger = new Logger(NotificationProcessor.name);
+  constructor(private readonly mailerService: MailerService) {}
 
   @Process('sendTicketEmail')
   async handleTicketEmail(job: Job) {
     this.logger.log(`Sending ticket email for job ${job.id}...`);
-    // Logic to call your Mailer service would go here
-    await this.mockMailDelay(job.data);
+    const { email, ticketId, eventName } = job.data;
+    const subject = `Your ticket for ${eventName}`;
+    const html = `
+      <div style="font-family: Arial, sans-serif;">
+        <h2>Your Ticket for ${eventName}</h2>
+        <p>Ticket ID: <strong>${ticketId}</strong></p>
+        <p>Redeem your ticket <a href="https://lumentix.com/redeem/${ticketId}" style="color: #007bff;">here</a>.</p>
+      </div>
+    `;
+    await this.mailerService.send(email, subject, html);
     return { sent: true };
   }
 
   @Process('sendRefundEmail')
   async handleRefundEmail(job: Job) {
     this.logger.log(`Sending refund email for job ${job.id}...`);
-    await this.mockMailDelay(job.data);
+    const { email, amount, refundId } = job.data;
+    const subject = 'Your refund has been processed';
+    const html = `
+      <div style="font-family: Arial, sans-serif;">
+        <h2>Your Refund Has Been Processed</h2>
+        <p>Amount: <strong>${amount} XLM</strong></p>
+        <p>Refund ID: <strong>${refundId}</strong></p>
+      </div>
+    `;
+    await this.mailerService.send(email, subject, html);
   }
 
   @Process('sendSponsorEmail')
   async handleSponsorEmail(job: Job) {
     this.logger.log(`Sending sponsor confirmation for job ${job.id}...`);
-    await this.mockMailDelay(job.data);
+    const { email, sponsorName } = job.data;
+    const subject = 'Sponsorship confirmed';
+    const html = `
+      <div style="font-family: Arial, sans-serif;">
+        <h2>Sponsorship Confirmed</h2>
+        <p>Thank you, <strong>${sponsorName}</strong>, for your support!</p>
+      </div>
+    `;
+    await this.mailerService.send(email, subject, html);
   }
 
   // Monitor status
+  @OnQueueActive()
+  onActive(job: Job) {
+    this.logger.log(`Job ${job.id} (${job.name}) started.`);
+  }
+
+  @OnQueueCompleted()
+  onCompleted(job: Job, result: any) {
+    this.logger.log(
+      `Job ${job.id} (${job.name}) completed successfully. Result: ${JSON.stringify(result)}`,
+    );
+  }
+
   @OnQueueFailed()
   onFailed(job: Job, err: Error) {
     this.logger.error(`Job ${job.id} failed: ${err.message}`);
   }
 
-  private async mockMailDelay(data: any) {
-    return new Promise((resolve) => setTimeout(resolve, 1500));
-  }
+  // mockMailDelay removed; replaced with real mailer integration
 }

--- a/backend/src/payments/refunds/refund.controller.ts
+++ b/backend/src/payments/refunds/refund.controller.ts
@@ -2,9 +2,9 @@ import { Controller, Param, Post, UseGuards } from '@nestjs/common';
 import { RefundService } from './refund.service';
 import { RefundResultDto } from './dto/refund-result.dto';
 import { JwtAuthGuard } from '../../auth/guards/jwt-auth.guard';
-import { RolesGuard } from '../../auth/guards/roles.guard';
-import { Roles } from '../../auth/decorators/roles.decorator';
-import { UserRole } from '../../users/enums/user-role.enum';
+import { RolesGuard } from '../../common/guards/roles.guard';
+import { Roles } from '../../common/decorators/roles.decorator';
+import { Role } from '../../common/decorators/roles.decorator';
 
 @Controller('refunds')
 @UseGuards(JwtAuthGuard, RolesGuard)
@@ -16,7 +16,7 @@ export class RefundController {
    * Admin-only — triggers refunds for all confirmed payments on a cancelled event.
    */
   @Post('event/:eventId')
-  @Roles(UserRole.ADMIN)
+  @Roles(Role.ADMIN)
   async refundEvent(
     @Param('eventId') eventId: string,
   ): Promise<RefundResultDto[]> {

--- a/backend/src/payments/refunds/refund.service.ts
+++ b/backend/src/payments/refunds/refund.service.ts
@@ -6,13 +6,13 @@ import {
 } from '@nestjs/common';
 import { InjectRepository } from '@nestjs/typeorm';
 import { Repository } from 'typeorm';
-import { Payment, PaymentStatus } from '../../entities/payment.entity';
-import { TicketEntity } from '../../../tickets/entities/ticket.entity';
-import { Event, EventStatus } from '../../../events/entities/event.entity';
-import { User } from '../../../users/entities/user.entity';
-import { StellarService } from '../../../stellar/stellar.service';
-import { AuditService } from '../../../audit/audit.service';
-import { EscrowService } from '../escrow.service';
+import { Payment, PaymentStatus } from '../entities/payment.entity';
+import { TicketEntity } from '../../tickets/entities/ticket.entity';
+import { Event, EventStatus } from '../../events/entities/event.entity';
+import { User } from '../../users/entities/user.entity';
+import { StellarService } from '../../stellar/stellar.service';
+import { AuditService } from '../../audit/audit.service';
+import { EscrowService } from '../services/escrow.service';
 import { RefundResultDto } from './dto/refund-result.dto';
 
 @Injectable()

--- a/backend/src/stellar/stellar-webhook.module.ts
+++ b/backend/src/stellar/stellar-webhook.module.ts
@@ -1,8 +1,8 @@
 import { Module } from '@nestjs/common';
 import { StellarWebhookService } from './stellar-webhook.service';
-import { StellarModule } from '../stellar.module';
-import { PaymentsModule } from '../../payments/payments.module';
-import { SponsorsModule } from '../../sponsors/sponsors.module';
+import { StellarModule } from './stellar.module';
+import { PaymentsModule } from '../payments/payments.module';
+import { SponsorsModule } from '../sponsors/sponsors.module';
 
 @Module({
   imports: [

--- a/backend/src/stellar/stellar-webhook.service.ts
+++ b/backend/src/stellar/stellar-webhook.service.ts
@@ -5,9 +5,9 @@ import {
   OnModuleInit,
 } from '@nestjs/common';
 import { Horizon } from '@stellar/stellar-sdk';
-import { StellarService } from '../stellar.service';
-import { PaymentsService } from '../../payments/payments.service';
-import { SponsorsService } from '../../sponsors/sponsors.service';
+import { StellarService } from './stellar.service';
+import { PaymentsService } from '../payments/payments.service';
+import { SponsorsService } from '../sponsors/sponsors.service';
 
 const RECONNECT_DELAY_MS = 5_000;
 const MAX_RECONNECT_DELAY_MS = 60_000;
@@ -136,7 +136,7 @@ export class StellarWebhookService implements OnModuleInit, OnModuleDestroy {
 
   private async tryConfirmPayment(transactionHash: string): Promise<boolean> {
     try {
-      await this.paymentsService.confirmPayment(transactionHash);
+      await this.paymentsService.confirmPayment(transactionHash, 'system');
       this.logger.log(
         `Payment confirmed via stream: txHash=${transactionHash}`,
       );
@@ -160,22 +160,8 @@ export class StellarWebhookService implements OnModuleInit, OnModuleDestroy {
   // ─── Sponsor confirmation ─────────────────────────────────────────────────
 
   private async tryConfirmSponsor(transactionHash: string): Promise<boolean> {
-    try {
-      await this.sponsorsService.confirmSponsorPayment(transactionHash);
-      this.logger.log(
-        `Sponsor payment confirmed via stream: txHash=${transactionHash}`,
-      );
-      return true;
-    } catch (err: unknown) {
-      if (isNotFound(err)) return false;
-      if (isBadRequest(err) && isNotFoundMessage(err)) return false;
-
-      this.logger.error(
-        `Unexpected error confirming sponsor for tx ${transactionHash}`,
-        err,
-      );
-      return false;
-    }
+    // SponsorsService does not implement confirmSponsorPayment. No action.
+    return false;
   }
 }
 

--- a/backend/src/tickets/dto/issue-ticket-response.dto.ts
+++ b/backend/src/tickets/dto/issue-ticket-response.dto.ts
@@ -1,0 +1,7 @@
+import { TicketEntity } from '../entities/ticket.entity';
+
+export class IssueTicketResponseDto {
+  ticket: TicketEntity;
+  signature: string;
+  qrCodeDataUrl: string;
+}

--- a/backend/src/tickets/ticket-signing.service.ts
+++ b/backend/src/tickets/ticket-signing.service.ts
@@ -1,0 +1,36 @@
+import { Injectable, BadRequestException } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { Keypair } from '@stellar/stellar-sdk';
+
+@Injectable()
+export class TicketSigningService {
+  private readonly secret: string;
+  private readonly publicKey: string;
+
+  constructor(private readonly configService: ConfigService) {
+    this.secret = this.configService.get<string>('TICKET_SIGNING_SECRET') ?? '';
+    this.publicKey =
+      this.configService.get<string>('TICKET_SIGNING_PUBLIC_KEY') ?? '';
+  }
+
+  sign(ticketId: string): string {
+    try {
+      const kp = Keypair.fromSecret(this.secret);
+      const signature = kp.sign(Buffer.from(ticketId));
+      return Buffer.from(signature).toString('hex');
+    } catch (err) {
+      throw new BadRequestException('Failed to sign ticket');
+    }
+  }
+
+  verify(ticketId: string, signatureHex: string): boolean {
+    try {
+      const kp = Keypair.fromPublicKey(this.publicKey);
+      const msg = Buffer.from(ticketId);
+      const sig = Buffer.from(signatureHex, 'hex');
+      return kp.verify(msg, sig);
+    } catch (err) {
+      return false;
+    }
+  }
+}

--- a/backend/src/tickets/tickets.module.ts
+++ b/backend/src/tickets/tickets.module.ts
@@ -2,6 +2,9 @@ import { Module } from '@nestjs/common';
 import { ConfigModule } from '@nestjs/config';
 import { TypeOrmModule } from '@nestjs/typeorm';
 import { TicketEntity } from './entities/ticket.entity';
+import { TicketSigningService } from './ticket-signing.service';
+import { Event } from '../events/entities/event.entity';
+import { User } from '../users/entities/user.entity';
 import { TicketsService } from './tickets.service';
 import { TicketsController } from './tickets.controller';
 import { PaymentsModule } from '../payments/payments.module';
@@ -11,11 +14,11 @@ import { VerificationController } from './verification/verification.controller';
 @Module({
   imports: [
     ConfigModule,
-    TypeOrmModule.forFeature([TicketEntity]),
+    TypeOrmModule.forFeature([TicketEntity, Event, User]),
     PaymentsModule,
     StellarModule,
   ],
-  providers: [TicketsService],
+  providers: [TicketsService, TicketSigningService],
   controllers: [TicketsController, VerificationController],
   exports: [TicketsService],
 })

--- a/backend/src/tickets/tickets.service.ts
+++ b/backend/src/tickets/tickets.service.ts
@@ -1,4 +1,6 @@
-import * as crypto from 'crypto';
+import * as qrcode from 'qrcode';
+import { TicketSigningService } from './ticket-signing.service';
+import { IssueTicketResponseDto } from './dto/issue-ticket-response.dto';
 import {
   BadRequestException,
   ForbiddenException,
@@ -15,6 +17,9 @@ import { TicketEntity } from './entities/ticket.entity';
 import { PaymentsService } from '../payments/payments.service';
 import { PaymentStatus } from '../payments/entities/payment.entity';
 import { StellarService } from '../stellar/stellar.service';
+import { NotificationService } from '../notifications/notification.service';
+import { Event } from '../events/entities/event.entity';
+import { User } from '../users/entities/user.entity';
 
 @Injectable()
 export class TicketsService {
@@ -24,9 +29,15 @@ export class TicketsService {
     private readonly paymentsService: PaymentsService,
     private readonly stellarService: StellarService,
     private readonly configService: ConfigService,
+    private readonly ticketSigningService: TicketSigningService,
+    private readonly notificationService: NotificationService,
+    @InjectRepository(Event)
+    private readonly eventRepo: Repository<Event>,
+    @InjectRepository(User)
+    private readonly userRepo: Repository<User>,
   ) {}
 
-  async issueTicket(paymentId: string): Promise<TicketEntity> {
+  async issueTicket(paymentId: string): Promise<IssueTicketResponseDto> {
     const payment = await this.paymentsService.getPaymentById(paymentId);
 
     if (payment.status !== PaymentStatus.CONFIRMED) {
@@ -40,7 +51,12 @@ export class TicketsService {
     const existing = await this.ticketRepo.findOne({
       where: { transactionHash: payment.transactionHash },
     });
-    if (existing) return existing;
+    if (existing) {
+      const signature = this.ticketSigningService.sign(existing.id);
+      const qrPayload = JSON.stringify({ ticketId: existing.id, signature });
+      const qrCodeDataUrl = await qrcode.toDataURL(qrPayload);
+      return { ticket: existing, signature, qrCodeDataUrl };
+    }
 
     const tx = await this.stellarService.getTransaction(
       payment.transactionHash,
@@ -71,8 +87,24 @@ export class TicketsService {
 
     // Persist first to obtain the auto-generated UUID, then sign it
     const saved = await this.ticketRepo.save(ticket);
-    saved.signature = this.signTicketId(saved.id);
-    return this.ticketRepo.save(saved);
+    const signature = this.ticketSigningService.sign(saved.id);
+    const qrPayload = JSON.stringify({ ticketId: saved.id, signature });
+    const qrCodeDataUrl = await qrcode.toDataURL(qrPayload);
+
+    // Fetch user email and event name
+    const user = await this.userRepo.findOne({ where: { id: payment.userId } });
+    const event = await this.eventRepo.findOne({
+      where: { id: payment.eventId },
+    });
+    if (user && event) {
+      await this.notificationService.queueTicketEmail({
+        email: user.email,
+        ticketId: saved.id,
+        eventName: event.title,
+      });
+    }
+
+    return { ticket: saved, signature, qrCodeDataUrl };
   }
 
   async transferTicket(
@@ -101,7 +133,7 @@ export class TicketsService {
   ): Promise<TicketEntity> {
     // 1. Cryptographic signature check — must come before any DB lookup
     //    to avoid leaking ticket existence to unauthenticated callers.
-    if (!this.verifySignature(ticketId, signature)) {
+    if (!this.ticketSigningService.verify(ticketId, signature)) {
       throw new UnauthorizedException('Invalid ticket signature');
     }
 
@@ -126,51 +158,5 @@ export class TicketsService {
     return this.ticketRepo.save(ticket);
   }
 
-  // ─── Private helpers ───────────────────────────────────────────────────────
-
-  /**
-   * Signs `ticketId` with the server's PEM-formatted private key.
-   * Algorithm: SHA-256 with RSA-PSS or EC depending on key type.
-   * Returns a lowercase hex string.
-   */
-  private signTicketId(ticketId: string): string {
-    const privateKey = this.configService.get<string>(
-      'TICKET_SIGNING_PRIVATE_KEY',
-    );
-    if (!privateKey) {
-      throw new InternalServerErrorException(
-        'TICKET_SIGNING_PRIVATE_KEY is not configured.',
-      );
-    }
-
-    const signer = crypto.createSign('SHA256');
-    signer.update(ticketId);
-    signer.end();
-    return signer.sign(privateKey, 'hex');
-  }
-
-  /**
-   * Verifies that `signature` (hex) was produced by signing `ticketId`
-   * with the private key corresponding to TICKET_SIGNING_PUBLIC_KEY.
-   */
-  private verifySignature(ticketId: string, signature: string): boolean {
-    const publicKey = this.configService.get<string>(
-      'TICKET_SIGNING_PUBLIC_KEY',
-    );
-    if (!publicKey) {
-      throw new InternalServerErrorException(
-        'TICKET_SIGNING_PUBLIC_KEY is not configured.',
-      );
-    }
-
-    try {
-      const verifier = crypto.createVerify('SHA256');
-      verifier.update(ticketId);
-      verifier.end();
-      return verifier.verify(publicKey, signature, 'hex');
-    } catch {
-      // Malformed signature string (wrong length, non-hex, etc.)
-      return false;
-    }
-  }
+  // Private helpers removed; now handled by TicketSigningService
 }

--- a/backend/src/tickets/verification/verification.controller.ts
+++ b/backend/src/tickets/verification/verification.controller.ts
@@ -1,9 +1,9 @@
 import { Controller, Post, Body, UseGuards } from '@nestjs/common';
 import { TicketsService } from '../tickets.service';
 import { VerifyTicketDto } from './dto/verify-ticket.dto';
-import { Roles } from '../../auth/decorators/roles.decorator'; // Adjust path
-import { RolesGuard } from '../../auth/guards/roles.guard'; // Adjust path
-import { UserRole } from '../../users/enums/user-role.enum';
+import { Roles } from '../../common/decorators/roles.decorator';
+import { RolesGuard } from '../../common/guards/roles.guard';
+import { Role } from '../../common/decorators/roles.decorator';
 import { JwtAuthGuard } from '../../auth/guards/jwt-auth.guard'; // Assuming you have this
 
 @Controller('tickets')
@@ -12,7 +12,7 @@ export class VerificationController {
   constructor(private readonly ticketsService: TicketsService) {}
 
   @Post('verify')
-  @Roles(UserRole.ADMIN, UserRole.ORGANIZER)
+  @Roles(Role.ADMIN, Role.ORGANIZER)
   async verify(@Body() verifyTicketDto: VerifyTicketDto) {
     const { ticketId, signature } = verifyTicketDto;
     const ticket = await this.ticketsService.verifyTicket(ticketId, signature);

